File: stores/juno.js
const fetch = require('node-fetch');
const cheerio = require('cheerio');

const STORE_NAME = 'Juno Download';
const STORE_URL = 'https://www.junodownload.com';

async function search(query) {
  try {
    const url = `https://www.junodownload.com/search/?q%5Ball%5D%5B%5D=${encodeURIComponent(query)}&solrorder=relevancy`;
    const res = await fetch(url, {
      timeout: 10000,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept': 'text/html,application/xhtml+xml',
        'Accept-Language': 'en-US,en;q=0.9',
      },
    });

    if (!res.ok) return [];

    const html = await res.text();
    const $ = cheerio.load(html);
    const results = [];

    $('.product-tracklist .product-tracklist-track, .jd-listing-item, .product').each((_, el) => {
      const $el = $(el);
      const title = $el.find('.product-tracklist-track-title a, .juno-title a, .product-title a').first().text().trim();
      const artist = $el.find('.product-tracklist-track-artists a, .juno-artist a, .product-artist a').map((_, a) => $(a).text().trim()).get().join(', ');
      const label = $el.find('.product-label a, .juno-label a').first().text().trim();
      const genre = $el.find('.product-genre a, .juno-genre a').first().text().trim();
      const price = $el.find('.product-buy .price, .buy-btn-price, .product-price').first().text().trim();
      const link = $el.find('.product-tracklist-track-title a, .juno-title a, .product-title a').first().attr('href');
      const img = $el.find('img').first().attr('data-src') || $el.find('img').first().attr('src');

      if (title) {
        results.push({
          title,
          artist,
          label,
          genre,
          bpm: null,
          key: null,
          duration: '',
          price: price || '£1.49',
          priceValue: parsePrice(price || '£1.49'),
          currency: 'GBP',
          artwork: img || null,
          url: link ? (link.startsWith('http') ? link : `${STORE_URL}${link}`) : `${STORE_URL}/search/?q%5Ball%5D%5B%5D=${encodeURIComponent(query)}`,
          store: STORE_NAME,
          storeIcon: 'juno',
          releaseDate: '',
        });
      }
    });

    return results.slice(0, 25);
  } catch (err) {
    console.error(`Juno Download search error: ${err.message}`);
    return [];
  }
}

function parsePrice(str) {
  const match = str.match(/[\d.]+/);
  return match ? parseFloat(match[0]) : null;
}

module.exports = { search, STORE_NAME };

File: stores/juno.py
from __future__ import annotations

import re
from urllib.parse import quote

import requests
from bs4 import BeautifulSoup

from models import TrackResult

STORE_NAME = "Juno Download"
_STORE_URL = "https://www.junodownload.com"
_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
    "Accept": "text/html,application/xhtml+xml",
    "Accept-Language": "en-US,en;q=0.9",
}

_PRICE_RE = re.compile(r"([£€$])\s*(\d+[.,]\d{2})")
_CURRENCY_BY_SYMBOL = {"£": "GBP", "€": "EUR", "$": "USD"}


def _parse_price(text: str):
    if not text:
        return text or "N/A", None, "GBP"
    m = _PRICE_RE.search(text)
    if not m:
        return text.strip(), None, "GBP"
    symbol, num = m.group(1), m.group(2).replace(",", ".")
    try:
        value = float(num)
    except ValueError:
        return text.strip(), None, "GBP"
    currency = _CURRENCY_BY_SYMBOL.get(symbol, "GBP")
    return f"{symbol}{value:.2f}", value, currency


def search(query: str) -> list[TrackResult]:
    try:
        url = f"{_STORE_URL}/search/?q%5Ball%5D%5B%5D={quote(query)}&solrorder=relevancy"
        resp = requests.get(url, headers=_HEADERS, timeout=10)
        if not resp.ok:
            return []

        soup = BeautifulSoup(resp.text, "html.parser")
        results: list[TrackResult] = []

        for el in soup.select(".jd-listing-item"):
            title_el = el.select_one(".juno-title")
            if not title_el:
                continue
            title = title_el.get_text(strip=True)
            link = title_el.get("href", "")
            if not title or not link:
                continue

            artist = ", ".join(a.get_text(strip=True) for a in el.select(".juno-artist a"))
            label_el = el.select_one(".juno-label")
            label = label_el.get_text(strip=True) if label_el else ""

            price_el = el.select_one("a.btn-cta")
            price_text = price_el.get_text(" ", strip=True) if price_el else ""
            price, price_value, currency = _parse_price(price_text)

            genre = ""
            release_date = ""
            meta_el = el.select_one(".text-muted")
            if meta_el:
                for line in [s.strip() for s in meta_el.stripped_strings]:
                    if re.match(r"^\d{1,2}\s+[A-Za-z]{3}\s+\d{2,4}$", line):
                        release_date = line
                    elif "/" in line or "House" in line or "Techno" in line or "Trance" in line:
                        genre = line

            img_el = el.select_one("img.li-img")
            img = None
            if img_el:
                img = img_el.get("src") or img_el.get("data-src")

            full_url = link if link.startswith("http") else f"{_STORE_URL}{link}"

            results.append(TrackResult(
                title=title,
                artist=artist,
                label=label,
                genre=genre,
                bpm=None,
                key=None,
                duration="",
                price=price,
                price_value=price_value,
                currency=currency,
                artwork=img,
                url=full_url,
                store=STORE_NAME,
                store_icon="juno",
                release_date=release_date,
            ))

            if len(results) >= 25:
                break

        return results
    except Exception as e:
        print(f"Juno Download search error: {e}")
        return []
